assembly/src/todo/mod.rs

```diff
@@ -0,0 +1,209 @@
+mod instruction;
+
+use core::borrow::Borrow;
+
+use vm_core::{code_blocks::CodeBlock, CodeBlockTable, Operation, Program};
+
+use crate::{
+    parsers::{self, Node, ProcedureAst, ProgramAst},
+    procedures::Procedure,
+    AssemblerError, ModuleProvider,
+};
+
+#[derive(Clone, Debug)]
+pub struct AssemblerContext {
+    cb_table: CodeBlockTable,
+    locals: Vec<Procedure>,
+}
+
+impl AssemblerContext {
+    pub fn new<I, P>(assembler: &Assembler, procs: I) -> Result<Self, AssemblerError>
+    where
+        I: Iterator<Item = P>,
+        P: Borrow<ProcedureAst>,
+    {
+        let mut context = Self {
+            cb_table: CodeBlockTable::default(),
+            locals: Vec::with_capacity(procs.size_hint().0),
+        };
+
+        for proc in procs {
+            let ProcedureAst {
+                name,
+                num_locals,
+                body,
+                is_export,
+                ..
+            } = proc.borrow();
+
+            let code_root = assembler.compile_body(&mut context, body.iter())?;
+            let proc = Procedure {
+                label: name.to_string(),
+                is_export: *is_export,
+                num_locals: *num_locals,
+                code_root,
+            };
+
+            context.locals.push(proc);
```

Comment:
Added with a pending question: what is the module path for local procedures? The module path is one of the arguments of the procedure id

Reply:
This will be needed for `CallLocal` instructions, right? I'll need to think more about this as we can't put MASTs for these procedures into procedure cache - so, something else might be needed.